Multiworkunit (bin packing) for kafka source

zliu41 opened the pull request:
- Implemented Kafka multiworkunit using bin packing algorithms. The goal is to evenly assign all workunits (each corresponding to a kafka partition) to the multiWorkUnits, while preferring to assign partitions of the same topic to the same multiWorkUnit (to avoid generating many small files).

*Note that currently it is not possible to avoid generating small files. Even if we assign multiple partitions of the same topic to the same multiWorkUnit, it will still create one file for each partition, rather than one file for the whole topic. It is currently neither possible to let multiple tasks share a writer (since tasks may run in parallel) nor possible to let each workUnit process multiple partitions (since each workUnit can only have one highwatermark). However, in the future if we change the design of watermarks such that it can hold multiple values, then a multiWorkUnit will be able to create one file for each topic, rather than each partition.*

The algorithm is slightly improved from Camus'. It first groups workUnits into approximately 3 * numOfMultiWorkunits groups using best-fit-decreasing, such that partitions of the same topic are in the same group (Camus uses an inferior algorithm for this step that may mix different topics in the same group). Then, assemble these groups into multiWorkunits using worst-fit-decreasing.

The first step uses best-fit-decreasing since the target is to group the partitions of each topic into as few groups as possible given a group size limit, so that partitions of the same topic are more likely to be assigned to the same mapper. Using a group size limit ensures that there are a sufficient number of groups for the second step, which is necessary for the second step to do an even assignment.

The reason behind 3* is that the worst-fit-decreasing algorithm should work well if the number of items is more than 3 times the number of bins.

The second step uses worst-fit-decreasing since the target is to evenly assign groups to multiWorkUnits.

- Changed the way Avro records from Kafka are written. Previously the AvroHdfsDataWriter calls setSchema() for each record whose schema is different from the current schema. This actually doesn't work since an avro file can only have one schema, and if we write records with different schemas to one file, the file cannot be decoded. Now KafkaAvroExtractor will convert the schema of a record if its schema is different from the latest schema of that topic.

lqiao commented:
Nice 2 level bin-packing. 

Is minimizing number of small file a goal or non-goal here? 

For the reduction of per partition files to be effective, after first level bin-packing (or second level to be better),  we need to create one work unit unifying the initial per topic/partition work units from a single WorkUnitGroup , where all the per topic/partition work units from that group are from the same topics. Otherwise, writer will create at least a file per work unit (per topic per partition) without unification, unless the writer behavior has changed.  

On the other hand, as we will have hourly compaction in place to merge small files into bigger ones, minimizing the number of small files will not be as critical as when we only have once in a day daily compaction (i.e. 24x small file # reduction).

zliu41 commented:
@lqiao Minimizing small files is not a critical goal, but since the 2-level approach should significantly reduce the number of small files without any noticeable impact on workunit skew (as long as number of groups > 3* number of mappers, bin packing will do a good job), I think we should still do it.

After first-level, each group (containing partitions of the same topic) will be assigned to a single multiWorkUnit (KafkaSource line 155).

zliu41 commented:
Since a topic usually has 10 to 100+ partitions and a mapper can usually be assigned at least 5 partitions, this should reduce the number of files by 5x or more. In my random tests the skew is never more than 3%.

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaExtractor.java:
```diff
@@ -41,6 +46,9 @@
 
   protected Iterator<MessageAndOffset> messageIterator;
   protected long nextWatermark;
+  private long totalRecordSize;
```
Is it possible to use `gobblin-metrics`, or this is what @ibuenros will be working on?

zliu41 replied:
It's better that @ibuenros replaces these with metrics as part of his ticket. There are other places in the kafka adapter that need to add metrics.

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -59,21 +64,28 @@
+        workUnits.add(getWorkUnitsForTopic(kafkaWrapper, topic, state));
       }
-      return builder.build();
+      int numOfMultiWorkunits = state.getPropAsInt(NUM_OF_MAPPERS, DEFAULT_NUM_OF_MAPPERS);
```
There's already a key named `mr.job.max.mappers` that defines the maximum number of mappers to launch. Can you use that?

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -83,6 +95,211 @@
+  private List<WorkUnit> getMultiWorkunits(List<List<WorkUnit>> workUnits, int numOfMultiWorkunits) {
+    Preconditions.checkArgument(numOfMultiWorkunits >= 1);
+
+    List<List<WorkUnitWithEstSize>> workUnitsWithEstSize = getWorkUnitsWithEstSize(workUnits);
```
Instead of creating this new class `WorkUnitWithEstSize`, why not simply setting a property in the `WorkUnit`?

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -83,6 +95,211 @@
+    long totalEstimatedDataSize = calcTotalEstSize(workUnitsWithEstSize);
+    long avgGroupSize = (long) ((double) totalEstimatedDataSize / (double) numOfMultiWorkunits / 3.0);
+
+    List<WorkUnitGroup> groups = Lists.newArrayList();
```
Similarly, `MultiWorkUnit` itself extends `WorkUnit` so you can set property in it and you don't need to create the new class `WorkUnitGroup`.

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -83,6 +95,211 @@
+   */
+  private static List<WorkUnit> worstFitDecreasingBinPacking(List<WorkUnitGroup> groups, int numOfMultiWorkUnits) {
+    List<MultiWorkUnit> multiWorkunits = Lists.newArrayListWithCapacity(numOfMultiWorkUnits);
+    List<Long> loads = Lists.newArrayListWithCapacity(numOfMultiWorkUnits);
```
Again, you can set the load as a property in the `MultiWorkUnit` so you don't need to maintain a separate list.

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -83,6 +95,211 @@
+    // Sort workunit groups by data size desc
+    Collections.sort(groups, sortBySizeDescComparator());
+
+    for (WorkUnitGroup group : groups) {
```
What if you use a `PriorityQueue` so the group with the minimum size is always at the top?

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -83,6 +95,211 @@
+    Collections.sort(workUnitsWithEstSize, sortBySizeDescComparator());
+
+    List<WorkUnitGroup> groups = Lists.newArrayList();
+    for (WorkUnitWithEstSize workUnitWithEstSize : workUnitsWithEstSize) {
```
Similarly, you can use a `PriorityQueue`.

liyinan926 commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -83,6 +95,211 @@
+   * A comparator for sorting workunits or workunit groups in descending order of estimated data size.
+   */
+  private static Comparator<WorkUnitWithEstSize> sortBySizeDescComparator() {
+    return new Comparator<WorkUnitWithEstSize>() {
```
This `Comparator` can be defined as a constant.

liyinan926 commented on gobblin-utility/src/main/java/gobblin/util/AvroUtils.java:
```diff
@@ -108,4 +123,29 @@
+  public static GenericRecord convertRecordSchema(GenericRecord record, Schema newSchema) throws IOException {
+    Preconditions.checkArgument(checkReaderWriterCompatibility(newSchema, record.getSchema()).getType() == COMPATIBLE);
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
```
Please make sure the output stream gets closed properly using a `Closer`.

zliu41 commented:
Updated (3a9b071).
- For bin packing, using priority queue instead of linear scan
- Removed class WorkUnitGroup and WorkUnitWithEstSize, now using MultiWorkUnit with a property "estimated.data.size".
- Other comments also addressed.

zliu41 commented:
Updated.

liyinan926 commented:
LGTM.
